### ISSUE TITLE
[CWM] Ongoing build and config model discussion

### ISSUE THREAD

**Issue by lin-d-hop** [DigitalCommons/mykomap#265]
*Track time under Coopeative World Map in Clockify*

## Description
Current MM has a lot of configuration that impacts both the build and how the live map is displayed to the user. There are complexities and problems with the current methods of doing this configuration that might be able to be simplified in the new MM.

Let's design the next version of the MM build and config.

## Acceptance Criteria
- A whiteboarding session to explore possibilities and agree design principles
- Document in this issue discussions and decisions about how to manage config throughout MM.
- Once the config design is clear, document in the new repo Wiki.

**Comment by lin-d-hop** [comment]
Notes from chat 20/8

https://docs.google.com/document/d/189uYra4YCMIe2yeguabDm_Qn8Th6wUCNvplyvLYrUVY/edit

TLDR;
Plan going forward
- Create monorepo (investigate best library to use for this, maybe pnpm), add mkyomap-react as a subpackage with git subtree, and add cwm-test-maps config.
- Create a build process which pulls in mykomap as a library and the cwm-test-maps config, and outputs an app.
- Create automated tests and CI pipeline which build and runs tests after each feature is committed, so that config and MM code stay in lock-step
- After we have got a working mykomap-react app with a backend that can consume actual data and config (by November? This feels quite late but we’re running out of time to create the actual CWM so we might have a lot of tech debt around this stuff post-November), try migrating an existing map (e.g. coopsuk) to the monorepo as a config folder. Then have another meeting to reassess our build + config architecture.
- Once our new architecture is confirmed, iteratively address the other pain points that were identified above.

**Comment by wu-lee** [comment]
Further to this, I have been doing some research and experimentation with monorepos / git subtrees.

## Monorepos

Some interesting overviews of the pros and cons of monorepos:

- DevOps Toolkit: [What Is A Monorepo And Why You Should Care](https://www.youtube.com/watch?v=VvcJGjjEyKo) - Best of the bunch I think, makes good points, esp on caveats and downsides
- Fireship: [Monorepos - How the Pros Scale Huge Softwre Projects // Turborepo vs Nx](https://www.youtube.com/watch?v=9iU_IE6vnJ8) - Describes why Lerner, Turborepo and Nx might be useful as a shortcut to avoid adopting the complex custom version  control and build  software used by Google et al. Useful if you think you want a monorepo for JS and and want to minimise the complexity.
- Software Developer Diaries: [I used a monorepo for 12 months - heres my opinion](https://www.youtube.com/watch?v=rcmdyQL2DUM) - Also interesting, lists pros and cons

Note Google (and MS, Facebook) use their own custom version control and build software to manage their use of gigantic monorepos, in an attempt to mitigate the problems (e.g. [they get super cumbersome][5]), but these are complex to use (["you'll need a PhD"][4]). And I get the strong impression we probably wouldn't want to do that.

As the second video above describes, there are other tools to help manage monorepos, e.g. Lerna, Turborepo and Nx, and these may offer some mitigations with less complexity, but nevertheless will still add some overhead for us to evaluate and then use.

## Git Subtree

None of these touch on git-subtree explicitly. For that maybe see the [manpage](https://manpages.debian.org/testing/git-man/git-subtree.1.en.html) and this [tutorial](https://medium.com/@v/git-subtrees-a-tutorial-6ff568381844). There is also this mediation on [git subtree split](https://medium.com/@jeevansathisocial/extract-subdirectory-from-git-repository-without-losing-history-3de8aed359a4). It seems you really need to go through the steps of creating one and see how it works in practice.

The difficulty I anticipate will be with managing branching and merging across multiple repos, which is where the alternative to git-subtree, git-submodules, really falls down in my experience.

How much better is git-subtree? I never used it so far because it looks a bit clunky too, and my inclination has been to avoid monorepos for components which need to be self-standing, using package management versioning to integrate the bits instead. (Which is what happens now with Mykomap). 

The boundary between these spaces (mono- vs poly-) is a bit permeable however - the data factory repo, `open-data` is a kind of mono-repo with lots of datasets wrapped into one. The machinery for the processing is in a separate repository however - `se-open-data`. The individual datasets can pick the version of it they want to use. I find the flexibility to do that is more useful than the cost of lack of uniformity, or perhaps I could say, I find the the cost of uniformity (dealing with breaking changes across all datasets at once) larger than I'd like to pay.

My experiment yesterday with git subtree suggests that everything's fine in the parent repo, but syncing work back to the child repos adds extra admin with some fairly long command-line steps, and requires an insight into what's actually being done on your behalf, which for me is still a bit fuzzy.

Basic things seem quite possible, like sending and receiving commits. Odd looking things do result, such as the commits you send get duplicated when you pull the changes back, but on another branch. Not a great side effect to have, but perhaps something we could work around.

A branch in the parent does *not* automatically result in a branch in the child repos, and vice versa, so I wonder if this will in practice put make branching and merging of child projects difficult and maybe skipped over if subtrees are used (although not as bad as git-submodules makes it). `git subtree split` is supposed to facilitate splitting off sub-directories as separate repos, preserving the history, but they don't in fact seem to preserve the branches and merges, just the commits in a flattened history. (git-repo-filter can do this, but is designed to be a one-off operation.) So the child repos seem like they might become essentially read-only single-branch views of commits in the parent, which means we're having cake but only eating the base layer.

### Git Subrepo

There is an alleged better alternative to subtree, [git subrepo](https://github.com/ingydotnet/git-subrepo). I gave that a quick go - it's not released as part of git like subtree currently is, but it's not hard to install. It's approach does seem somewhat nicer, perhaps - more sophisticated, and simpler to use.  But again I wonder about the implications for workflow, which still seems like it gets more complicated. And I don't know how well supported it might be in the future - other tools like this such as git-slave seem to have rusted by the wayside some years ago, and we wouldn't want to be a passenger in another one.

## Summary

Overall, I think my main concern is that there might indeed be some benefit to investigating these things, but perhaps it's unwise to invite too much change right now? We have a lot on our plate as it is. Just researching and evaluating these things takes time and implementing them might be distracting and cumbersome, at least to start with.

I am thinking maybe we should use the KISS principle and avoid trying to make big choices about things which are not on the critical path to getting the results we hope for in the time available.  We don't want to force ourselves into hasty decisions nor to get sidetracked by trying to evaluate the possibilities. [Brooks' Law][2] seems relevant - he says that adding new team members to a project under pressure just makes it later. A corollary might be that changing their processes could have an equivalent effect. (Unless perhaps there's a demonstrably low-risk and simple to execute change that can accelerate progress in a short enough window, which I'm not sure that coming up with a new build process, branching model and repository organisation really is.)

The bare essential I think we have not *not* yet answered is whether it is realistically possible to build a 500k map which can load and filter pins fast enough with the available technology, yet alone have feature parity with the existing Mykomap. Yet we have promised to deliver that. I'm uncomfortable about this.

At the moment, we have shown we can pan and zoom a MapLibre application with 500k pins generated in-browser, but for me, that seems to take several seconds to load the data and show any pins at all.

### So?

This is not entirely contradicting all of the above. 

To try and suggest a practical way forward, can we keep working on a "spike" project (or projects) to show it seems feasible to do all the things we want to do? Use Dave Thomas' idea of ["tracer bullets"][1] - a prototype which implements all the essentials end-to-end, but can be tweaked until it visibly hits the target. 

The details therefore probably to be allowed to be fairly ad-hoc and variable, so that alternative implementations can be compared - in case it's not so easy to get it right. We can try and spec things out but not necessarily in lots of detail.  e.g I've written a Swagger spec before, I'll see if I can draft one for #267 quickly.

Then we can flesh that working prototype out, whilst keeping track of our benchmarks (which we also need to define) - when we see the parameters we care about going off target, we know which component(s) were responsible.

Obviously keeping an eye to being able to use this for the other Mykomap use-cases, but I suggest the priorities should be:
1. Flag any problems early - and manage client expectations should we need to
2. Deliver something which works for the CWM project
3. Make it flexible enough to use for the other pre-existing Mykomap cases (probably after delivery, unless we're lucky to have made lots of progress)

This could be done in a single repo, perhaps multiple will be useful. We could have a "trunk-driven" branching process, but we won't be maintaining releases yet, so it's going to be features branches off a dev branch anyway. I don't think that's very important right now. Nor are the build processes, so long as they work without too much hassle. I think we could get by with what we have more or less, and not try and spend time researching and evaluating alternatives.

Choice of software platforms might be more important, as they're hard to change later. But I think the main question left now is what goes on the back-end. (To think about this briefly, I would suggest something which works smoothly with Swagger/OpenAPI and is small and simple and aimed at rapid prototyping. I see that [NocoDB says it supports Swagger][3], and we can run this on our Cloudron server, which simplifies deployment massively)

[1]: https://www.artima.com/articles/tracer-bullets-and-prototypes
[2]: https://en.wikipedia.org/wiki/Brooks%27s_law
[3]: https://docs.nocodb.com/0.109.7/developer-resources/accessing-apis/
[4]: https://youtu.be/9iU_IE6vnJ8?t=127
[5]: https://youtu.be/VvcJGjjEyKo?t=446

**Comment by rogup** [comment]
@wu-lee I agree with the idea of treating this CWM project as a 'spike' since we don't know for sure how easy/possible it will be to integrate the existing maps into the new way of doing things, and we don't have the time to test this out fully. I think as part of this spike, we should take the opportunity to take some risks and try new technologies and solutions that address existing pain points. And we should be bold when making these decisions, since we don't have time to ruminate over every decision or keep all our options open indefinitely, but be willing to quickly react if things don't work and try a different option. 

If we follow KISS principles, the minimum requirement is that this new project works for the CWM by the end of Nov. And in general, a monorepo feels simpler than creating multiple new repos for different parts of the project. We don't need to worry about transferring existing maps to the new structure at this stage.

Then after the project, we'll have tech debt to transfer the old maps, and maybe change some of the ways that we implemented things if needed. But it feels like a waste to be overly-cautious in this spike and default to reimplementing old ways of doing things that knowingly cause problems, when we're going through the effort of starting a fresh codebase. Either way we'll be left with tech debt. We can just try to mindful of how existing maps would eventually migrate to this new system.

I have some comments on some specific technical points that you made:

- Is there an actual requirement to publish individual subrepos if we use a monorepo? We're currently the only ones who use the MM codebase, and Lynne said in the last meeting that it's unlikely anyone else will. If someone does want to reuse our libraries, it's not much effort for them to pull the monorepo and split out the library themselves. Or we could configure Git Subtree at a later time, if someone asks for it. It sounds like your investigation shows there are solutions out there that will probably be good enough, if it comes to it.
- > The bare essential I think we have not not yet answered is whether it is realistically possible to build a 500k map which can load and filter pins fast enough with the available technology, yet alone have feature parity with the existing Mykomap. Yet we have promised to deliver that. I'm uncomfortable about this. At the moment, we have shown we can pan and zoom a MapLibre application with 500k pins generated in-browser, but for me, that seems to take several seconds to load the data and show any pins at all.

  I thought this is answered now? In #263 I've shown that filtering is pretty much instant (https://digitalcommons.github.io/mykomap-react/), and the demo with using vector tiles shows that the initial load can be very fast https://dev.maps.solidarityeconomy.coop/experiment/cwm-test-maps/ I don't think we need to do any more investigations or demos at this stage, I feel fairly confident in our proposed solution and we just need to have a go at making it.
- > We could have a "trunk-driven" branching process, but we won't be maintaining releases yet, so it's going to be features branches off a dev branch anyway.
  
  The point of the trunk-driven branching model is that we don't have feature branches off a dev, we have a main branch that we commit everything to (or very short-lived branches where we break up larger features into parts across multiple PRs if needed). I think the idea is to avoid big merge conflicts when merging multiple conflicting features, which is probably what we'll be doing. We've just been working off main branch in https://github.com/DigitalCommons/mykomap-react so far, so I'd be up for trying to continue this, and reassess if it's not working.

- > But I think the main question left now is what goes on the back-end. (To think about this briefly, I would suggest something which works smoothly with Swagger/OpenAPI and is small and simple and aimed at rapid prototyping. I see that [NocoDB says it supports Swagger](https://docs.nocodb.com/0.109.7/developer-resources/accessing-apis/), and we can run this on our Cloudron server, which simplifies deployment massively)
  
  We need to schedule our whiteboard session (#266) to decide this stuff.

**Comment by rogup** [comment]
I'm blocked on everything atm because we're waiting on whiteboards or we haven't fully committed to decisions collectively. But we have a lot of work to do this week and it feels like we're already getting behind on stuff.

So I think I'm just going to go ahead with the next steps that we planned in the whiteboard (outlined by Lynne in this comment https://github.com/DigitalCommons/mykomap/issues/265#issuecomment-2299210511) and see how it goes

**Comment by rogup** [comment]
> Create monorepo (investigate best library to use for this, maybe pnpm), add mkyomap-react as a subpackage with git subtree, and add cwm-test-maps config.

Lots of articles and forums online recommend pnpm workspaces as a good dependency manager for monorepos. We can also use Nx/Turborepo on top of this, but this can be added later. These libraries are used to streamline the build process and are probably more useful for larger teams that have long, complicated workflows, so that build artefacts can be cached in a distributed way.

**Comment by wu-lee** [comment]
I've had a chat with @rogup in a stand-up call earlier, and we discussed the above issues, so I'll just try and sum up quickly here.

> And we should be bold when making these decisions, since we don't have time to ruminate over every decision or keep all our options open indefinitely, but be willing to quickly react if things don't work and try a different option.

Exactly, I agree, so long as we don't try and boldy take on lots of things we don't need to. I don't mind putting stuff into one repo for now if that seems to help, just don't want to be stuck with this indefinitely. And I don't really care about the branching details - mainly I want us to avoid the need to go and read lots of guides on how to implement some new scheme, or to have to struggle through the initial learning curve of new tools.

>     * Is there an actual requirement to publish individual subrepos if we use a monorepo? We're currently the only ones who use the MM codebase, and Lynne said in the last meeting that it's unlikely anyone else will. If someone does want to reuse our libraries, it's not much effort for them to pull the monorepo and split out the library themselves. Or we could configure Git Subtree at a later time, if someone asks for it. It sounds like your investigation shows there are solutions out there that will probably be good enough, if it comes to it.

Although I'm totally convinced mono-repos are the panacea for our ills, for now I don't care about avoiding them nor about publishing subrepos, and if we're not stuck with this situation am happy for us to stuff a bunch of things in one repository if this seems to help. 

>     * > The bare essential I think we have not not yet answered is whether it is realistically possible to build a 500k map which can load and filter pins fast enough with the available technology, yet alone have feature parity with the existing Mykomap. Yet we have promised to deliver that. I'm uncomfortable about this. At the moment, we have shown we can pan and zoom a MapLibre application with 500k pins generated in-browser, but for me, that seems to take several seconds to load the data and show any pins at all.
>       
>       
>       I thought this is answered now? In [[CWM] First iteration MM with MapLibre #263](https://github.com/DigitalCommons/mykomap/issues/263) I've shown that filtering is pretty much instant (https://digitalcommons.github.io/mykomap-react/), and the demo with using vector tiles shows that the initial load can be very fast https://dev.maps.solidarityeconomy.coop/experiment/cwm-test-maps/ I don't think we need to do any more investigations or demos at this stage, I feel fairly confident in our proposed solution and we just need to have a go at making it.

Ah. I had not seen that comment in the other issue. My GitHub notifications come thick and fast and so tend to pile up. Don't rely on me seeing them quickly.  

So this is good, although I gather it's only been tested on your laptop, which I think is a bit too high spec to demonstrate we've got a working solution for most people, especially people on mobile hardware.  I'll have a look at how it does on my 12-year old desktop.

As performance degrades is non-linearly, and it can sometimes just take a little extra to move from "adequate" to "thrashing" (perhaps because of garbage collection), I think we also need to represent all the parts of the chain, including loading data from a server and converting that into the format needed.

>     * > We could have a "trunk-driven" branching process, but we won't be maintaining releases yet, so it's going to be features branches off a dev branch anyway.
>       
>       
>       The point of the trunk-driven branching model is that we don't have feature branches off a dev, we have a main branch that we commit everything to (or very short-lived branches where we break up larger features into parts across multiple PRs if needed). I think the idea is to avoid big merge conflicts when merging multiple conflicting features, which is probably what we'll be doing. We've just been working off main branch in https://github.com/DigitalCommons/mykomap-react so far, so I'd be up for trying to continue this, and reassess if it's not working.

Yep. I think I tend to aim to do this (small feature branches) anyway, as well as small commits. 

The question of whether we have one long-lived branch or two (dev and master in Mykomap) is kind of academic at the moment, as we will only have one in practice at this point, whichever choice we make.
 
The main caveat is that we likely need to be able to define a config variant for each "environment" (development, production) for our applications *in the same branch*, and select one based on some runtime parameter supplied in the deployed system, otherwise we get the same problem as we do in Mykomap currently, which then leads to the branching scheme used as a workaround.

>     * > But I think the main question left now is what goes on the back-end. (To think about this briefly, I would suggest something which works smoothly with Swagger/OpenAPI and is small and simple and aimed at rapid prototyping. I see that [NocoDB says it supports Swagger](https://docs.nocodb.com/0.109.7/developer-resources/accessing-apis/), and we can run this on our Cloudron server, which simplifies deployment massively)
>       
>       
>       We need to schedule our whiteboard session ([[CWM] Whiteboarding session for creating the MM backend #266](https://github.com/DigitalCommons/mykomap/issues/266)) to decide this stuff.

We had this chat earlier. More details in thread here https://github.com/DigitalCommons/mykomap/issues/266

**Comment by rogup** [comment]
I've implemented a mono-repo here https://github.com/DigitalCommons/mykomap-monolith

I used `git-filter-repo` to move the `mykomap-react` repo into a subfolder whilst preserving the history. And set up Github pages to still deploy the front-end at https://digitalcommons.github.io/mykomap-monolith/

I've kept everything as simple as possible, so we're still using `npm`, but with their `workspaces` feature. It's very simple to use, see the root `package.json` and `package-lock.json` to see how dependencies are installed and scripts are run for each workspace. And there is a bit more info here.
https://docs.npmjs.com/cli/v9/using-npm/workspaces

I think most of the discussion is resolved in this ticket and we've roughly converged on a path to try out for the CWM project. I'll keep this ticket in review though since we'll need to refer back to it and probably change plans, as work progresses and new ideas emerge.
